### PR TITLE
feat: offline with git submodule

### PR DESCRIPTION
**Pull request by QiChenX**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
feature

* **What is the current behavior?** (You can also link to an open issue here)
https://github.com/4paradigm/OpenMLDB/issues/3760
manage rapidjson directly using git submodule

* **What is the new behavior (if this is a feature change)?**
bundle rapidjson & farmhash directly in source code, allowing users to compile OpenMLDB offline

### CODE REVIEW

**Comment by aceforeverd** [comment]
requires #3778

File: docker/Dockerfile_offline
Comment (aceforeverd) [inline docker/Dockerfile_offline:1]: `FROM ghcr.io/4paradigm/hybridsql:{version}`

not really meaningful to write twice

File: Makefile [hunk 1]
```diff
@@ -124,7 +124,10 @@ test: build
 	$(CMAKE_PRG) --build $(OPENMLDB_BUILD_DIR) --target test -- -j$(NPROC)
 	sh ./steps/ut_zookeeper.sh stop
 
+# .git must exists, otherwise git submodule will fail
 configure: thirdparty-fast
+	git submodule init
```
Comment (aceforeverd) [inline Makefile]: not necessary after #3778

File: CMakeLists.txt [hunk 2]
```diff
@@ -281,6 +279,7 @@ include_directories(
   ${LLVM_INCLUDE_DIRS}
   ${Boost_INCLUDE_DIRS}
   ${CMAKE_PREFIX_PATH}/include
+  ${PROJECT_SOURCE_DIR}/contrib/rapidjson/include
```
Comment (aceforeverd) [inline CMakeLists.txt:297]: ~~use `target_include_directories` in rapidjson directory~~

Not sure, is it possible to add the include directory in relative CMakeLists.txt from `contrib/rapidjson-cmake` directory ?

File: .gitmodules [hunk 3]
```diff
@@ -1,3 +1,9 @@
 [submodule "cmake-modules"]
 	path = cmake-modules
 	url = https://github.com/bilke/cmake-modules.git
+[submodule "contrib/farmhash"]
```
Comment (aceforeverd) [inline .gitmodules]: I think we can move farmhash's source code into our code base, it is archived, and just two relative source

File: contrib/CMakeLists.txt [hunk 4]
```diff
@@ -26,3 +26,19 @@ endfunction()
 
 add_contrib(simdjson)
 
+# farmhash
+add_library(farmhash ${CMAKE_CURRENT_SOURCE_DIR}/farmhash/src/farmhash.cc)
```
Comment (aceforeverd) [inline contrib/CMakeLists.txt]: a separate CMakeLists.txt required

File: contrib/rapidjson
Comment (dl239) [inline contrib/rapidjson:1]: set a fixed tag version

File: .gitmodules [hunk 5]
```diff
@@ -1,3 +1,9 @@
...
+	path = contrib/farmhash
+	url = https://github.com/google/farmhash.git
+[submodule "contrib/rapidjson"]
```
Comment (dl239) [inline .gitmodules:10]: why add  rapidjson to submodule

File: docker/Dockerfile_offline [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    yum clean all && rm -v lcov-1.15-1.noarch.rpm && \
+    curl -Lo zookeeper.tar.gz https://archive.apache.org/dist/zookeeper/zookeeper-3.4.14/zookeeper-3.4.14.tar.gz && \
+    mkdir -p /deps/src && \
+    tar xzf zookeeper.tar.gz -C /deps/src && \
+    rm -v ./*.tar.gz && \
+    /setup_deps.sh -a "$TARGETARCH" -z "$ZETASQL_VERSION" -t "$THIRDPARTY_VERSION" && \
+    rm -v /setup_deps.sh
+
+# install rsync
+RUN yum install -y rsync
```
Comment (dl239) [inline docker/Dockerfile_offline]: install together with the above installing commands

File: farmhash/src/farmhash.h
Comment (aceforeverd) [inline farmhash/src/farmhash.h:1]: we only care about the two source code file,  do not add any other file from farmhash repo

File: docker/Dockerfile [hunk 7]
```diff
@@ -33,7 +33,8 @@ RUN yum update -y && yum install -y centos-release-scl epel-release && \
     tar xzf zookeeper.tar.gz -C /deps/src && \
     rm -v ./*.tar.gz && \
     /setup_deps.sh -a "$TARGETARCH" -z "$ZETASQL_VERSION" -t "$THIRDPARTY_VERSION" && \
-    rm -v /setup_deps.sh
```
Comment (aceforeverd) [inline docker/Dockerfile]: why this ?

File: docker/Dockerfile_offline [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Before building an offline image, copy this Dockerfile to the same directory as OpenMLDB firstly
+
+# specify a base version of OpenMLDB docker image
+ARG version=0.8
```
Comment (aceforeverd) [inline docker/Dockerfile_offline]: use `0.8.5`

File: docker/Dockerfile_offline [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+FROM ghcr.io/4paradigm/hybridsql:${version}
+
+COPY OpenMLDB/docker/setup_deps.sh /
+COPY OpenMLDB /root/OpenMLDB
```
Comment (aceforeverd) [inline docker/Dockerfile_offline]: use git clone

File: docker/Dockerfile_offline [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Before building an offline image, copy this Dockerfile to the same directory as OpenMLDB firstly
+
+# specify a base version of OpenMLDB docker image
+ARG version=0.8
+FROM ghcr.io/4paradigm/hybridsql:${version}
+
+COPY OpenMLDB/docker/setup_deps.sh /
```
Comment (aceforeverd) [inline docker/Dockerfile_offline]: never used

File: docker/Dockerfile_offline [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+COPY OpenMLDB /root/OpenMLDB
+
+# current base image does not contain rsync
+RUN yum install -y rsync
```
Comment (aceforeverd) [inline docker/Dockerfile_offline]: always do a `yum clean all` afterwards

File: docker/Dockerfile [hunk 12]
```diff
@@ -33,7 +33,8 @@ RUN yum update -y && yum install -y centos-release-scl epel-release && \
...
+    rm -v /setup_deps.sh && \
+    yum install -y rsync
```
Comment (aceforeverd) [inline docker/Dockerfile]: I prefer the compile image as minimal, rsync is not required dep based on current workflow

Comment (QiChenX) [reply docker/Dockerfile]: maybe we can keep rsync in Dockerfile_offline and remove it in Dockerfile?

Comment (tobegit3hub) [reply docker/Dockerfile]: It is okay to delete this since we need to connect Internet to deploy.

File: farmhash/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,12 @@
+
```
Comment (tobegit3hub) [inline farmhash/CMakeLists.txt]: Since we add this as submodule, we may remove the source code.

File: docker/Dockerfile_offline [hunk 14]
```diff
@@ -0,0 +1,25 @@
+# Copyright 2021 4Paradigm
```
Comment (tobegit3hub) [inline docker/Dockerfile_offline]: We can rename this as "openmldb-build.dockerfile".

File: farmhash/CMakeLists.txt (hunk 13 above)
Comment (QiChenX) [reply farmhash/CMakeLists.txt]: now we directly include the required source code of farmhash in our code base

File: farmhash/CMakeLists.txt
Comment (aceforeverd) [inline farmhash/CMakeLists.txt:1]: pls follow the guide, vendered source should inside contrib directory

File: docker/Dockerfile (hunk 7 above)
Comment (aceforeverd) [reply docker/Dockerfile]: I mean this `Dockerfile` apparently does not requires change

File: docker/Dockerfile_offline (hunk 14 above)
Comment (aceforeverd) [reply docker/Dockerfile_offline]: `full-offline.Dockerfile` still necessary, or it confuse to the original Dockerfile

File: docker/openmldb-build.dockerfile [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+# limitations under the License.
+
+# Before building an offline image, copy this Dockerfile to the same directory as OpenMLDB firstly
+
+# specify a base version of OpenMLDB docker image
+ARG version=0.8.5
+FROM ghcr.io/4paradigm/hybridsql:${version}
+
+RUN cd /root && \
+    git clone https://github.com/4paradigm/OpenMLDB.git && \
```
Comment (aceforeverd) [inline docker/openmldb-build.dockerfile]: with submodules

File: CMakeLists.txt [hunk 16]
```diff
@@ -145,8 +145,6 @@ endif()
 # third party sources
 include(FetchContent)
```
Comment (aceforeverd) [inline CMakeLists.txt]: rm this FetchContent module

**Review by aceforeverd — CHANGES_REQUESTED** [review]

**Comment by aceforeverd** [comment]
@QiChenX pls ensure all CIs pass

File: contrib/farmhash/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,12 @@
+
+add_library(farmhash ${CMAKE_CURRENT_SOURCE_DIR}/src/farmhash.cc)
+target_include_directories(farmhash SYSTEM PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/src/")
+set_target_properties(farmhash
+    PROPERTIES
+    LIBRARY_OUTPUT_DIRECTORY ${farmhash_BINARY_DIR}
```
Comment (aceforeverd) [inline contrib/farmhash/CMakeLists.txt]: now the `framhash_BINARY_DIR` goes undefined, please update to CMAKE_CURRENT_BINARY_DIR

u should verify where the built library outputs

**Review by aceforeverd — APPROVED** [review]

**Review by dl239 — APPROVED** [review]

**Pull request merged**